swagger_server/dal/sample.py

```diff
@@ -0,0 +1,99 @@
+from neobolt.exceptions import ConstraintError
+
+from swagger_server.helpers import db
+from swagger_server.models import Sample, Neighbour, NearestLeaf
+
+
+class SampleAlreadyExist(Exception):
+    pass
+
+
+class SampleNotExist(Exception):
+    pass
+
+
```

Comment:
There are a lot going on in this class. It looks to me the responsibilities of this class include those that are usually found in the service layer and the repository layer. I will put the business logic in the service layer and leave the repositories for pure db queries. For example, I would recommend implement a sample repository like this:

```suggestion
def find_sample_by_id(id) -> Sample
def delete_sample_by_id(id)
def find_nearest_neighbours_by_id(id) -> List(Neighbour)
def add_nearest_neighbours_by_id(id, List(Neighbour)) -> List(Neighbour)
def delete_nearest_neighbours_by_id(id)
...
```

Reply:
What would the business logic / service layer look like?

Reply:
I usually consider that an endpoint implementation can be served by three layers working together:

- a thin controller layer whose responsibilities include http request/response handling
- a service layer which is responsible for anything in between, makes interfaces available to controllers, uses repository interfaces for data CRUDs, validates user input, handle db query results, applies business rules to the data etc. If this is a complex application, then this layer could be further partitioned.
- a thin repository layer which implements the repository interfaces that service layer uses. This may or may not include ORM etc. depending on the implementation.